Add aqara lock support

SchumyHao opened the pull request:
Signed-off-by: SchumyHao <schumyhaojl@126.com>

## Description:
Add aqara lock support

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

homeassistant commented:
Hello @SchumyHao,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`35c0466414c149ce9b894b205359e8ee3aba3cb1`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;14419&#x2F;commits&#x2F;35c0466414c149ce9b894b205359e8ee3aba3cb1) - This commit has something that looks like an email address (bob-hjl@126.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hi @SchumyHao, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#14419).

Once you do that we will be able to review and accept this pull request.

Thanks!

syssi commented:
Plesse fix the lint errors. I'm very happy with your approach!

syssi commented:
I've release PyXiaomiGateway 0.9.2. Please update the requirements!

SchumyHao commented:
Hi @syssi 
I have update the requirement.
Thank you for your review.

MartinHjelmare commented on homeassistant/components/lock/xiaomi_aqara.py:
```diff
@@ -0,0 +1,70 @@
+"""Support for Xiaomi Gateway Lock."""
```
Add a link to the docs page as in other platforms.

MartinHjelmare commented on homeassistant/components/lock/xiaomi_aqara.py:
```diff
@@ -0,0 +1,70 @@
+    """Perform the setup for Xiaomi devices."""
+    devices = []
+
+    for (_, gateway) in hass.data[PY_XIAOMI_GATEWAY].gateways.items():
```
```py
for gateway in hass.data[PY_XIAOMI_GATEWAY].gateways.values():
...
```

syssi commented:
requirements_all.txt is not up to date
Please run script/gen_requirements_all.py

MartinHjelmare commented on homeassistant/components/lock/xiaomi_aqara.py:
```diff
@@ -0,0 +1,70 @@
+        self._changed_by = 0
+        self._verified_wrong_times = 0
+
+        XiaomiDevice.__init__(self, device, name, xiaomi_hub)
```
```py
super().__init__(device, name, xiaomi_hub)
```

MartinHjelmare commented on homeassistant/components/lock/xiaomi_aqara.py:
```diff
@@ -0,0 +1,70 @@
+    def device_state_attributes(self) -> dict:
+        """Return the state attributes."""
+        attributes = {}
+        attributes[ATTR_VERIFIED_WRONG_TIMES] = self._verified_wrong_times
```
```py
attributes = {
    ATTR_VERIFIED_WRONG_TIMES: self._verified_wrong_times,
}
```

SchumyHao committed:
xiaomi_aqara: add aqara lock support

Signed-off-by: SchumyHao <schumyhaojl@126.com>

MartinHjelmare commented on homeassistant/components/lock/xiaomi_aqara.py:
```diff
@@ -0,0 +1,70 @@
+    @property
+    def is_locked(self) -> bool:
+        """Return true if lock is locked."""
+        return True
```
So this entity doesn't show the current state nor can change the state of the lock. What's the idea for this platform? How should it be used?

SchumyHao replied:
Xiaomi Aqara lock can not know the door's status.
This lock only can told user verified_id when someone unlock the lock.

MartinHjelmare replied:
Is there any precedence for accepting a platform in home assistant that doesn't change state and doesn't expose any services?

MartinHjelmare replied:
Yes. But the automation will only trigger if someone else or something else changed the lock than the previous changer. If this use case is good enough, I'm fine with this. But you should make it clear in the docs what the limitations of the platform are.

syssi replied:
The `last_updated` attribute will help to identify this use case.

SchumyHao replied:
I have push a patch to change lock state automatically
After verity OK, the lock will be unlock state for about 5s.
Then, the lock go back unlock state automatically.

SchumyHao committed:
fix some format issue

Signed-off-by: SchumyHao <schumyhaojl@126.com>

syssi commented on homeassistant/components/lock/xiaomi_aqara.py:
```diff
@@ -1,4 +1,9 @@
-"""Support for Xiaomi Gateway Lock."""
+"""
+Support for Xiaomi Gateway Lock.
```
Could you call it Xiaomi Aqara Lock or similar. I don't like the "Gateway" here.

syssi commented:
Please update (PyXiaomiGateway=0.9.3) or rebase the PR. Thanks in advance! :-)

SchumyHao committed:
xiaomi_aqara: change Gateway to Aqara

Signed-off-by: SchumyHao <schumyhaojl@126.com>

SchumyHao committed:
xiaomi_aqara: change lock state

Aqara lock can not read lock state.
But After verify successfully, lock will be unlock state
for about 5s.
After 5s, lock will auto reset to lock state.

Signed-off-by: SchumyHao <schumyhaojl@126.com>

syssi reviewed (approved):
LGTM.

syssi commented on homeassistant/components/lock/xiaomi_aqara.py:
```diff
@@ -0,0 +1,91 @@
+                self._changed_by = int(value)
+                self._verified_wrong_times = 0
+                self._state = STATE_UNLOCKED
+                async_call_later(self.hass, UNLOCK_MAINTAIN_TIME_S,
```
Please remove the unit (`_S`).

MartinHjelmare commented on homeassistant/components/lock/xiaomi_aqara.py:
```diff
@@ -0,0 +1,91 @@
+UNLOCK_MAINTAIN_TIME_S = 5
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
This should be made a coroutine via `async def` to make the call to `async_call_later` in `parse_data`, which will be called during entity init, work. I assume the rest of the calls in the base entity `XiaomiDevice` can be done safely from an async context?

SchumyHao committed:
xiaomi_aqara: remove UNLOCK_MAINTAIN_TIME unit

Signed-off-by: SchumyHao <schumyhaojl@126.com>

SchumyHao committed:
xiaomi_aqara: setup_platform in coroutine

Signed-off-by: SchumyHao <schumyhaojl@126.com>

syssi commented on homeassistant/components/lock/xiaomi_aqara.py:
```diff
@@ -0,0 +1,91 @@
+        }
+        return attributes
+
+    @asyncio.coroutine
```
Could you care about this one, too?

SchumyHao replied:
I'm not sure for this comment.
This function will be invoked in async_call_later, I need this function called after UNLOCK_MAINTAIN_TIME seconds. but callback function can not do IO operation, I think delay should not called in callback too.

MartinHjelmare replied:
`async_call_later` is a callback.

SchumyHao replied:
I try to change this by using async_run_job, I tested this can work:
```
@@ -67,8 +68,10 @@ class XiaomiAqaraLock(LockDevice, XiaomiDevice):
         return attributes
 
     @asyncio.coroutine
-    def clear_unlock_state(self, _):
+    def clear_unlock_state(self):
         """Clear unlock state automatically."""
+        yield from asyncio.sleep(UNLOCK_MAINTAIN_TIME)
+
         self._state = STATE_LOCKED
         self.async_schedule_update_ha_state()
 
@@ -85,8 +88,7 @@ class XiaomiAqaraLock(LockDevice, XiaomiDevice):
                 self._changed_by = int(value)
                 self._verified_wrong_times = 0
                 self._state = STATE_UNLOCKED
-                async_call_later(self.hass, UNLOCK_MAINTAIN_TIME,
-                                 self.clear_unlock_state)
+                self.hass.async_run_job(self.clear_unlock_state)
                 return True
 
         return False

```

But I failed to using *@callback*
I have tried this 
```
@@ -66,9 +67,12 @@ class XiaomiAqaraLock(LockDevice, XiaomiDevice):
         }
         return attributes
 
+    @callback
     @asyncio.coroutine
-    def clear_unlock_state(self, _):
+    def clear_unlock_state(self):
         """Clear unlock state automatically."""
+        yield from asyncio.sleep(UNLOCK_MAINTAIN_TIME)
+
         self._state = STATE_LOCKED
         self.async_schedule_update_ha_state()
 
@@ -85,8 +89,7 @@ class XiaomiAqaraLock(LockDevice, XiaomiDevice):
                 self._changed_by = int(value)
                 self._verified_wrong_times = 0
                 self._state = STATE_UNLOCKED
-                async_call_later(self.hass, UNLOCK_MAINTAIN_TIME,
-                                 self.clear_unlock_state)
+                self.hass.async_run_job(self.clear_unlock_state)
                 return True
 
```
But ```clear_unlock_state(self)``` will not invoked.

I'm sorry that how to use callback decorator, can you point me out how to do this?

SchumyHao replied:
Hi @MartinHjelmare  @syssi 
May you help me to improve this code?
I have checked the code in homeassistant, which used async_call_later and callback, but do not find a similar usage in my situation.

SchumyHao replied:
Done.
Thank you for your advice

SchumyHao committed:
xiaomi_aqara: use callback instead of asyncio.coroutine

Signed-off-by: SchumyHao <schumyhaojl@126.com>

syssi commented:
Could you add some docs?

SchumyHao commented:
OK, I'll do this job.
I have forked https://github.com/SchumyHao/home-assistant.github.io I'll create a PR later.

syssi commented:
I added some basic docs: https://github.com/home-assistant/home-assistant.github.io/pull/5387

SchumyHao commented:
I'm not sure what should I do in next step?
@syssi  Can you show me what should I do to push this PR merged?

SchumyHao commented:
Thank you @syssi. Please let me know if I need do something more.
